comms/src/control_service/worker.rs

```diff
@@ -187,82 +189,278 @@ where
-                warn!(target: LOG_TARGET, "Underlying socket closed on '{}'", event.address);
-            },
-            _ => {},
+                self.reject_connection(envelope_header, identity_frame)?;
+                return Ok(());
+            }
         }
 
-        Ok(())
-    }
+        if let Some(conn) = conn_manager.get_connection(&peer) {
+            if conn.is_active() {
+                debug!(
+                    target: LOG_TARGET,
+                    "Already have active connection to peer. Ignoring the request for connection."
+                );
+                self.reject_connection(envelope_header, identity_frame)?;
+                return Ok(());
+            }
+        }
+
+        self.accept_connection(envelope_header, identity_frame)?;
+
+        debug!(
+            target: LOG_TARGET,
+            "Connecting to requested address {}", message.address
+        );
 
-    fn establish_listener(&self) -> Result<EstablishedConnection> {
+        let conn = conn_manager
+            .establish_requested_outbound_connection(&peer, message.address.clone(), message.server_key)
+            .map_err(ControlServiceError::ConnectionManagerError)?;
+
+        conn.wait_connected_or_failure(&self.config.requested_outbound_connection_timeout)
+            .map_err(ControlServiceError::ConnectionError)?;
         debug!(
             target: LOG_TARGET,
-            "Binding on address: {}", self.config.listener_address
+            "Connection to requested address {} succeeded", message.address
         );
-        Connection::new(&self.context, Direction::Inbound)
+
+        self.send_message_to_peer(&conn, &peer, self.config.accept_message_type.clone(), Accept {})?;
+
+        debug!(
+            target: LOG_TARGET,
+            "Sent 'Accept' message to address {:?}",
+            conn.get_connected_address()
+        );
+
+        Ok(())
+    }
+
+    fn should_reject_collision(&self, node_id: &NodeId) -> bool {
+        &self.node_identity.identity.node_id < node_id
+    }
+
+    fn reject_connection(&self, envelope_header: MessageEnvelopeHeader, identity: Frame) -> Result<()> {
+        self.send_reply(
+            &envelope_header.source,
+            identity,
+            ControlServiceMessageType::ConnectRequestOutcome,
+            ConnectRequestOutcome::Rejected,
+        )
+    }
+
+    fn accept_connection(&self, envelope_header: MessageEnvelopeHeader, identity: Frame) -> Result<()> {
+        self.send_reply(
+            &envelope_header.source,
+            identity,
+            ControlServiceMessageType::ConnectRequestOutcome,
+            ConnectRequestOutcome::Accepted,
+        )
+    }
+
+    fn get_peer(&self, public_key: &CommsPublicKey) -> Result<Option<Peer>> {
+        let peer_manager = &self.connection_manager.peer_manager();
+        match peer_manager.find_with_public_key(public_key) {
+            Ok(peer) => Ok(Some(peer)),
+            Err(PeerManagerError::PeerNotFoundError) => Ok(None),
```

Comment:
The PeerManager should probably be changed to rather return a None and not a PeerNotFoundError error.